### PR TITLE
Benchmarks: Add benchmark: Megatron-LM/Megatron-Deepspeed GPT pretrain benchmark

### PR DESCRIPTION
**Pull request by yukirora**
**Description**
Megatron-LM/Megatron-Deepspeed GPT pretrain benchmark

### CODE REVIEW

File: third_party/Makefile [hunk 1]
```diff
@@ -171,3 +171,20 @@ directx_amf_encoding_latency:
 		del vs_buildtools.exe && echo "Deleted vs_buildtools.exe" && \
 		"C:\temp\BuildTools\MSBuild\Current\Bin\MSBuild.exe" "AMF\amf\public\samples\CPPSamples_vs2019.sln" /t:EncoderLatency /p:Platform=x64 /p:Configuration=Release /p:OutDir="%SB_MICRO_PATH%\bin" \
 	)
+
+# Install Megatron-LM
+megatron_lm:
```
Comment (cp5555) [inline third_party/Makefile:176]: What's for AMD support? Is there easy to change from Megatron-DeepSpeed to Megatron-LM?

File: superbench/benchmarks/model_benchmarks/model_base.py [hunk 2]
```diff
@@ -263,6 +264,10 @@ def __train(self, precision):
 
         # The unit of step time should be millisecond.
         step_times = self._train_step(precision)
+        if isinstance(step_times, tuple):
+            info = step_times[1]
+            step_times = step_times[0]
```
Comment (cp5555) [inline superbench/benchmarks/model_benchmarks/model_base.py:269]: step_times = step_times[0]
info = step_times[1]

File: superbench/benchmarks/model_benchmarks/model_base.py [hunk 3]
```diff
@@ -468,3 +474,13 @@ def _log_step_time(self, curr_step, precision, duration):
             step_time = statistics.mean(duration) if len(duration) < self._args.log_n_steps \
                 else statistics.mean(duration[-self._args.log_n_steps:])
             stdout_logger.log(f'{self._name} - {precision.value}: step {curr_step}, step time {step_time}\n')
+
+    def _process_other_info(self, model_action, precision, info):
```
Comment (cp5555) [inline superbench/benchmarks/model_benchmarks/model_base.py:478]: rename to "_process_info()"

File: superbench/benchmarks/model_benchmarks/megatron_gpt3.py [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+            data (list): the data to be reduced.
+        """
+        comm = MPI.COMM_WORLD
+        data = np.array(data, dtype=np.float64)
+        # Reduce the data to a single value on rank 0
+        result = np.zeros_like(data)
+        comm.Allreduce([data, MPI.DOUBLE], [result, MPI.DOUBLE], op=MPI.MAX)
+        return result.tolist()
+
+    def _process_other_info(self, model_action, precision, info):
```
Comment (cp5555) [inline superbench/benchmarks/model_benchmarks/megatron_gpt3.py:353]: _process_info()

File: superbench/benchmarks/model_benchmarks/megatron_gpt3.py [hunk 5]
```diff
@@ -0,0 +1,509 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (cp5555) [inline superbench/benchmarks/model_benchmarks/megatron_gpt3.py:1]: Do we need to add in default for how to use it?

File: docs/user-tutorial/benchmarks/model-benchmarks.md [hunk 6]
```diff
@@ -42,3 +42,24 @@ For inference, supported percentiles include
...
+
+#### Metrics
+| Name                                              | Unit                   | Description                                                             |
+|---------------------------------------------------|------------------------|-------------------------------------------------------------------------|
+| megatron-gpt/${precision}_train_step_time         | time (ms)              | The average training step time between each whole global batch.         |
+| megatron-gpt/${precision}_train_throughput        | throughput (samples/s) | The average training throughput between each whole global batch.        |
+| megatron-gpt/${precision}_train_tflops            | tflops/s               | The average training tflops per second between each whole global batch. |
+| megatron-gpt/${precision}_train_mem_allocated     | GB                     | The average GPU memory allocated between each whole global batch.       |
+| megatron-gpt/${precision}_train_max_mem_allocated | GB                     | The average max GPU memory allocated between each whole global batch.   |
```
Comment (cp5555) [inline docs/user-tutorial/benchmarks/model-benchmarks.md:64]: What's the meaning of "between each whole global batch"

#### Metrics
| Name                                              | Unit                   | Description                                                             |
|---------------------------------------------------|------------------------|-------------------------------------------------------------------------|
| megatron-gpt/${precision}_train_step_time         | time (ms)              | The average training step time per iteration.         |
| megatron-gpt/${precision}_train_throughput        | throughput (samples/s) | The average training throughput per iteration.        |
| megatron-gpt/${precision}_train_tflops            | tflops/s               | The average training tflops per second iteration. |
| megatron-gpt/${precision}_train_mem_allocated     | GB                     | The average GPU memory allocated per iteration  .       |
| megatron-gpt/${precision}_train_max_mem_allocated | GB                     | The average maximum GPU memory allocated per iteration.  |

File: docs/user-tutorial/benchmarks/model-benchmarks.md [hunk 7]
```diff
@@ -42,3 +42,24 @@ For inference, supported percentiles include
 | model-benchmarks/pytorch-${model_name}/${precision}_inference_throughput                | throughput (samples/s) | The average inference throughput with fp32/fp16 precision.                   |
```
Comment (cp5555) [inline docs/user-tutorial/benchmarks/model-benchmarks.md:42]: Whether we need to change training throughput? It should be training throughput per GPU?

**Review by cp5555 — APPROVED** [review]

File: docs/user-tutorial/benchmarks/model-benchmarks.md (hunk 6 above)
Comment (guoshzhao) [reply docs/user-tutorial/benchmarks/model-benchmarks.md:64]: Does here means 'per whole global batch'?

**Review by guoshzhao — APPROVED** [review]

**Pull request merged**